objective.py

```diff
@@ -26,29 +29,79 @@ def set_data(self, X, y):
         self.X, self.y = X, y
         self.n_samples, self.n_features = X.shape
 
+        # the following code fits the path once using the same path-stopping
+        # criteria that glmnet uses to find a reasonable lambda path in order
+        # to obtain a reasonable lambda sequence
         if self.lambda_min_ratio is None:
             self.lambda_min_ratio = 1e-2 if self.n_samples < self.n_features else 1e-4
 
         lambda_max = self._get_lambda_max()
-        self.lambdas = np.logspace(
+        lambdas = np.logspace(
             np.log(lambda_max),
             np.log(lambda_max * self.lambda_min_ratio),
             num=self.n_lambda,
             base=np.exp(1),
         )
 
+        # the pmax setting is redundant because dfmax is hardcoded, but it's
+        # kept just to be explicit about how this is done
+        dfmax = self.n_features + 1
+        pmax = min(dfmax * 2 + 20, self.n_features)
+
+        lasso = Lasso(
+            max_iter=10000,
+            alpha=lambda_max / self.n_samples,
+            # TODO(jolars): consider whether tol might be too high
+            tol=1e-4,
+            fit_intercept=self.fit_intercept,
+            warm_start=True,
+        )
+
+        r = y - np.mean(y) if self.fit_intercept else y.copy()
+        null_dev = 0.5 * np.linalg.norm(r) ** 2
+        dev_prev = null_dev
+
+        i = 0
+        ever_active = np.array([])
+
+        for i, lam in enumerate(lambdas):
+            lasso.alpha = lam / self.n_samples
+            lasso.fit(X, y)
+
+            w = lasso.coef_
+            intercept = lasso.intercept_
+
+            dev = 0.5 * np.linalg.norm(y - X @ w - intercept) ** 2
+
+            dev_ratio = 1 - dev / null_dev
+            dev_change = 1 - dev / dev_prev
+            dev_prev = dev
+
+            n_nonzero = np.sum(w != 0)
+            ever_active = np.union1d(ever_active, np.where(w != 0))
+
+            if (
+                n_nonzero >= dfmax
+                or dev_ratio > 0.999
+                or (i > 0 and dev_change < 1e-5)
+                or len(ever_active) >= pmax
+            ):
+                break
+
+        self.lambdas = lambdas[: i + 1]
```

Comment:
no space after : (?)

Reply:
> LGTM, minor comments

Thanks for the review and useful comments!